## Investigasi Sampel Data Titanic, berikut dengan cara :
1. Cek secara head, tail, sample, info lalu observasi apa yang bisa anda peroleh ... ?
2. Lakukan Statistical Summary dengan mengekstrak informasi yang didapat dari observasi anda ... ?
3. Cek apakah ada duplikat dan bagaimana handlenya ... ?
4. Cek apakah ada missing value, berapa persentasenya jika ada, dan bagaimana cara handlenya ... ?


---



Penungasan Dikerjakan oleh **Haidar Nabiilah Sunu** dari *SMK Telkom Purwokerto* ...

## Import Libraries

In [ ]:
# Mengimpor library numpy untuk operasi numerik
import numpy as np
# Mengimpor library pandas untuk manipulasi data
import pandas as pd
# Mengimpor seaborn untuk visualisasi data yang menarik dan informatif
import seaborn as sns
# Mengimpor matplotlib.pyplot untuk membuat grafik dan plot
import matplotlib.pyplot as plt
# Mengatur tampilan pandas agar menampilkan semua kolom saat mencetak DataFrame
pd.set_option("display.max_columns", None)
# Mengatur tampilan pandas agar menampilkan semua baris saat mencetak DataFrame
pd.set_option("display.max_rows", None)

## Check Handle of the Head, Tail, dan Sample for Data

In [ ]:
# Import data
df = pd.read_excel('titanic.xlsx')
df.head() # Import data head [ Menampilan 5 data teratas dari 'titanic.xlsx' ]

survived                                             name     sex      age
0         1                    Allen, Miss. Elisabeth Walton  female  29.0000
1         1                   Allison, Master. Hudson Trevor    male   0.9167
2         0                     Allison, Miss. Helen Loraine  female   2.0000
3         0             Allison, Mr. Hudson Joshua Creighton    male  30.0000
4         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.0000

In [ ]:
# Import data tail [ Menampilkan 5 data terbawah dari 'titanic.xlsx' ]
df.tail()

survived                                     name     sex   age
495         1  Mallet, Mrs. Albert (Antoinette Magnin)  female  24.0
496         0        Mangiavacchi, Mr. Serafino Emilio    male   NaN
497         0               Matthews, Mr. William John    male  30.0
498         0                Maybery, Mr. Frank Hubert    male  40.0
499         0                McCrae, Mr. Arthur Gordon    male  32.0

In [ ]:
# Import data sample 5 baris data [ diberi secara acak tertiap dijalankan ]
df.sample(5)

survived                           name     sex   age
0           1  Allen, Miss. Elisabeth Walton  female  29.0
9           0        Artagaveytia, Mr. Ramon    male  71.0
268         0       Smith, Mr. Lucien Philip    male  24.0
184         0              Lewy, Mr. Ervin G    male   NaN
30          0   Blackwell, Mr. Stephen Weart    male  45.0

Pengamatan :
1. Semua kolom (survived, name, sex, age) terlihat sesuai dengan tipe data yang diharapkan: survived berupa nilai biner (0 atau 1), name berupa string, sex berupa kategori (male/female), dan age berupa numerik.

2. Terdapat nilai NaN pada kolom age, yang menunjukkan adanya missing data dan perlu dipertimbangkan untuk ditangani (misalnya dengan imputasi).

3. Nilai pada kolom sex hanya terdiri dari dua kategori, menandakan kolom ini bisa dikonversi menjadi data kategorikal/biner.

4. Tidak ditemukan anomali yang mencolok antara nama kolom dan isi datanya — semuanya tampak logis dan sesuai konteks.

In [ ]:
# Menampilkan informasi ringkas tentang struktur DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  500 non-null    int64  
 1   name      500 non-null    object 
 2   sex       500 non-null    object 
 3   age       451 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 15.8+ KB


Pengamatan :
1. Dataset terdiri dari 500 baris dan 4 kolom: survived, name, sex, dan age.

2. Tiga kolom (survived, name, dan sex) tidak memiliki nilai yang hilang (non-null = 500), sedangkan kolom age memiliki 451 nilai, berarti ada 49 data yang hilang.

3. Tipe data sudah sesuai: survived bertipe integer (karena biner 0/1), name dan sex bertipe objek (string/kategori), dan age bertipe float karena mungkin ada nilai desimal atau nilai kosong yang menyebabkan konversi otomatis.

4. Penggunaan memori relatif kecil, hanya sekitar 15.8 KB, menandakan data cukup ringan untuk diproses.

## Statistical Summary

In [ ]:
# Menampilakan Kolom data
df.columns

Index(['survived', 'name', 'sex', 'age'], dtype='object')

In [ ]:
# Membuat kelompok 'categoricals' dan 'numericals'
categoricals = ['sex'] # Menentukan kolom-kolom kategorikal

# Menentukan kolom-kolom numerikal
numericals = ['survived', 'name', 'age',]

In [ ]:
# Ringkasan statistik Numerik
df[numericals].describe()

survived         age
count  500.000000  451.000000
mean     0.540000   35.917775
std      0.498897   14.766454
min      0.000000    0.666700
25%      0.000000   24.000000
50%      1.000000   35.000000
75%      1.000000   47.000000
max      1.000000   80.000000

Pengamatan :
* Nilai minimum dan maksimum pada kolom survived dan age terlihat masuk akal. Kolom survived memiliki nilai 0 dan 1, menandakan data biner (tidak selamat atau selamat), sedangkan kolom age berkisar dari bayi (0.67 tahun) hingga lansia (80 tahun).

* Rata-rata usia (35.9) mendekati median (35), menunjukkan bahwa distribusi usia cukup simetris.

* Kolom survived adalah data biner, sehingga tidak perlu dianalisis simetrisitasnya. Yang perlu diperhatikan hanya keseimbangan kelas : 54% selamat dan 46% tidak selamat, menunjukkan distribusi kelas yang cukup seimbang.

* Untuk analisis lebih lanjut, distribusi nilai pada kolom-kolom lainnya dapat divisualisasikan, khususnya jika ada kolom kategorikal atau diskrit lainnya.

In [ ]:
# Menampilkan statistik deskriptif untuk kolom kategorikal
df[categoricals].describe()

sex
count    500
unique     2
top     male
freq     288

Pengamatan :
* Kolom sex memiliki total 500 entri tanpa missing value.

* Terdapat 2 nilai unik dalam kolom ini, yaitu male dan female.

* Nilai yang paling sering muncul adalah male dengan frekuensi 288, menunjukkan bahwa mayoritas individu dalam data adalah laki-laki.

In [ ]:
categoricals

['sex']

In [ ]:
for col in categoricals:
  print(f"Value counts of {col} column")
  print(df[col].value_counts(), '\n')

Value counts of sex column
sex
male      288
female    212
Name: count, dtype: int64 



In [ ]:
for col in numericals:
  print(f"==== {col} ====")
  print(df[col].value_counts(), '\n')

==== survived ====
survived
1    270
0    230
Name: count, dtype: int64 

==== name ====
name
Eustis, Miss. Elizabeth Mussey                                                        2
Becker, Miss. Ruth Elizabeth                                                          1
Becker, Miss. Marion Louise                                                           1
Becker, Master. Richard F                                                             1
Beauchamp, Mr. Henry James                                                            1
Beane, Mrs. Edward (Ethel Clarke)                                                     1
Beane, Mr. Edward                                                                     1
Bateman, Rev. Robert James                                                            1
Banfield, Mr. Frederick James                                                         1
Ball, Mrs. (Ada E Hall)                                                               1
Baimbrigge, Mr. Charles Ro

# Data Inspection and Cleaning




## 1. Duplicate Handling
### Identifying Duplicate


In [ ]:
len(df) # Menampilan jumlah baris di data

500

In [ ]:
# Menampilkan jumlah baris setelah menghapus duplikat
len(df.drop_duplicates())

499

In [ ]:
# Menghitung proporsi baris unik dibandingkan dengan total baris
len(df.drop_duplicates()) / len(df)

0.998

In [ ]:
list(df.columns) # Mendapatkan daftar kolom dalam DataFrame

['survived', 'name', 'sex', 'age']

In [ ]:
# Ambil baris duplikat (termasuk yang asli)
duplicates = df[df.duplicated(keep=False)]

In [ ]:
duplicates

survived                            name     sex   age
104         1  Eustis, Miss. Elizabeth Mussey  female  54.0
349         1  Eustis, Miss. Elizabeth Mussey  female  54.0

In [ ]:
duplicates.groupby(list(df.columns)).size()

survived  name                            sex     age 
1         Eustis, Miss. Elizabeth Mussey  female  54.0    2
dtype: int64

In [ ]:
duplicates.groupby(list(df.columns)).size().reset_index(name='jumlah duplikat')

survived                            name     sex   age  jumlah duplikat
0         1  Eustis, Miss. Elizabeth Mussey  female  54.0                2

In [ ]:
# Hitung frekuensi kemunculan tiap baris duplikat
duplicate_counts = duplicates.groupby(list(df.columns)).size().reset_index(name='jumlah_duplikat')

# Urutkan berdasarkan jumlah duplikat
sorted_duplicates = duplicate_counts.sort_values(by='jumlah_duplikat', ascending=False)

# Tampilkan hasil
print("Baris duplikat yang sudah diurutkan berdasarkan jumlah kemunculannya:")
sorted_duplicates

Baris duplikat yang sudah diurutkan berdasarkan jumlah kemunculannya:


survived                            name     sex   age  jumlah_duplikat
0         1  Eustis, Miss. Elizabeth Mussey  female  54.0                2

Pada kode (df[df.duplicated(keep=False)]), Anda langsung mencetak baris yang memiliki duplikat dengan menggunakan metode duplicated dan memasukkan parameter keep=False. Ini mencetak semua baris yang memiliki duplikat tanpa mengurangi apa pun dari DataFrame.

In [ ]:
# Menghapus duplikat dan menyimpan hasilnya kembali ke DataFrame (df)
df = df.drop_duplicates()

In [ ]:
len(df.drop_duplicates()) / len(df)
# Jika output dari code di cell ini tidak bernilai 1 maka terdapat duplikat

1.0

#### Telah di Handling Drop Duplikat dari Data

## 2. Missing Value Handling
### Identifying Missing Value

In [ ]:
# Mengecek jumlah nilai yang hilang (NaN) di setiap kolom DataFrame
df.isna().sum()

survived     0
name         0
sex          0
age         49
dtype: int64

In [ ]:
# Mengembalikan DataFrame dengan nilai yang mengandung NaN (null)
df.isnull().sum()

survived     0
name         0
sex          0
age         49
dtype: int64

In [ ]:
for column in df.columns:
    print(f"============= {column} =================")
    display(df[column].value_counts())
    print()

============= survived =================


survived
1    269
0    230
Name: count, dtype: int64


============= name =================


name
McCrae, Mr. Arthur Gordon                                                             1
Allen, Miss. Elisabeth Walton                                                         1
Allison, Master. Hudson Trevor                                                        1
Lehmann, Miss. Bertha                                                                 1
Laroche, Mrs. Joseph (Juliette Marie Louise Lafargue)                                 1
Laroche, Mr. Joseph Philippe Lemercier                                                1
Laroche, Miss. Simonne Marie Anne Andree                                              1
Laroche, Miss. Louise                                                                 1
Lamb, Mr. John Joseph                                                                 1
Lahtinen, Rev. William                                                                1
Lahtinen, Mrs. William (Anna Sylfven)                                                 1
Kvillner, Mr. Johan Henrik Johannesson                                                1
Knight, Mr. Robert J                                                                  1
Kirkland, Rev. Charles Leonard                                                        1
Kelly, Mrs. Florence "Fannie"                                                         1
Keane, Mr. Daniel                                                                     1
Keane, Miss. Nora A                                                                   1
Karnes, Mrs. J Frank (Claire Bennett)                                                 1
Kantor, Mrs. Sinai (Miriam Sternin)                                                   1
Borebank, Mr. John James                                                              1
Bonnell, Miss. Elizabeth                                                              1
Bonnell, Miss. Caroline                                                               1
Blank, Mr. Henry                                                                      1
Blackwell, Mr. Stephen Weart                                                          1
Bjornstrom-Steffansson, Mr. Mauritz Hakan                                             1
Bissette, Miss. Amelia                                                                1
Bishop, Mrs. Dickinson H (Helen Walton)                                               1
Bishop, Mr. Dickinson H                                                               1
Birnbaum, Mr. Jakob                                                                   1
Bird, Miss. Ellen                                                                     1
Bidois, Miss. Rosalie                                                                 1
Behr, Mr. Karl Howell                                                                 1
Beckwith, Mrs. Richard Leonard (Sallie Monypeny)                                      1
Beckwith, Mr. Richard Leonard                                                         1
Beattie, Mr. Thomson                                                                  1
Cardeza, Mrs. James Warburton Martinez (Charlotte Wardle Drake)                       1
Cardeza, Mr. Thomas Drake Martinez                                                    1
Candee, Mrs. Edward (Helen Churchill Hungerford)                                      1
Calderhead, Mr. Edward Pennington                                                     1
Cairns, Mr. Alexander                                                                 1
Butt, Major. Archibald Willingham                                                     1
Burns, Miss. Elizabeth Margaret                                                       1
Bucknell, Mrs. William Robert (Emma Eliza Ward)                                       1
Brown, Mrs. John Murray (Caroline Lane Lamson)                                        1
Brown, Mrs. James Joseph (Margaret Tobin)                                             1
Brewe, Dr. Arthur Jackson          


============= sex =================


sex
male      288
female    211
Name: count, dtype: int64


============= age =================


age
24.0000    23
30.0000    20
36.0000    19
18.0000    14
42.0000    14
45.0000    14
35.0000    14
22.0000    12
28.0000    12
23.0000    11
21.0000    11
31.0000    11
48.0000    11
25.0000    11
29.0000    10
34.0000    10
50.0000    10
47.0000    10
39.0000     9
27.0000     9
54.0000     9
33.0000     8
49.0000     8
38.0000     8
19.0000     8
40.0000     8
55.0000     8
32.0000     7
60.0000     7
17.0000     7
26.0000     7
44.0000     6
52.0000     6
37.0000     6
58.0000     6
64.0000     5
46.0000     5
57.0000     5
51.0000     5
16.0000     4
53.0000     4
56.0000     4
43.0000     4
62.0000     4
61.0000     4
1.0000      3
8.0000      3
63.0000     3
41.0000     3
15.0000     2
71.0000     2
20.0000     2
65.0000     2
4.0000      2
6.0000      2
2.0000      1
0.9167      1
14.0000     1
11.0000     1
76.0000     1
59.0000     1
80.0000     1
28.5000     1
32.5000     1
70.0000     1
45.5000     1
13.0000     1
67.0000     1
12.0000     1
0.8333      1
0.6667      1
7.0000      1
3.0000      1
Name: count, dtype: int64

In [ ]:
# Menghitung jumlah baris (row) dalam DataFrame (df)
total_rows = len(df)
total_rows # Menampilkan nilai dari variabel total_rows

499

In [ ]:
# Menampilakan Kolom data
df.columns

Index(['survived', 'name', 'sex', 'age'], dtype='object')

In [ ]:
# Menghitung total baris dalam DataFrame
total_rows = len(df)

# Loop untuk setiap kolom di DataFrame
for column in df.columns:
    # Hitung jumlah nilai yang hilang (NaN) di kolom
    missing_count = df[column].isna().sum()

    # Hitung persentase nilai yang hilang terhadap total baris
    missing_percentage = (missing_count / total_rows) * 100

    # Tampilkan jumlah dan persentase nilai yang hilang
    print(f"Column '{column}' Has {missing_count} missing values ({missing_percentage:.2f}%)")

Column 'survived' Has 0 missing values (0.00%)
Column 'name' Has 0 missing values (0.00%)
Column 'sex' Has 0 missing values (0.00%)
Column 'age' Has 49 missing values (9.82%)


Pengamatan :
* Kolom 'age' memiliki 49 missing values (sekitar 9.82%).

* Kolom lain seperti 'survived', 'name', dan 'sex' tidak memiliki nilai hilang (0%).

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 499 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  499 non-null    int64  
 1   name      499 non-null    object 
 2   sex       499 non-null    object 
 3   age       450 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 19.5+ KB


In [ ]:
df['sex'].dtype

dtype('O')

In [ ]:
df['sex'].mode()[0]

'male'

In [ ]:
df['survived'].median()

1.0

In [ ]:
# Loop untuk setiap kolom dalam DataFrame
for column in df.columns:

    # Jika tipe datanya adalah object (kategorikal)
    if df[column].dtype == 'object':
        # Isi missing value dengan nilai paling sering (modus)
        df[column].fillna(df[column].mode()[0], inplace=True)

    else:
        # Jika numerik, isi missing value dengan nilai tengah (median)
        df[column].fillna(df[column].median(), inplace=True)

<ipython-input-88-46c7b10ce65e>:11: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[column].fillna(df[column].median(), inplace=True)
<ipython-input-88-46c7b10ce65e>:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try

In [ ]:
df.isna().sum()

survived    0
name        0
sex         0
age         0
dtype: int64

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 499 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  499 non-null    int64  
 1   name      499 non-null    object 
 2   sex       499 non-null    object 
 3   age       499 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 19.5+ KB


Berhasil menghapus nilai NULL pada DataFrame ... !

## Thank you :)